File: spike/routing_recovery.py
"""Pure routing recovery model with no sockets, files, or OS side effects."""

from dataclasses import dataclass


ROUTE_LOCAL_BYPASS = "local_bypass"
ROUTE_GEO_EXIT = "geo_exit"
ROUTE_UNKNOWN = "unknown"

SERVICE_DISCORD = "discord"
SERVICE_YOUTUBE = "youtube_video"

POLICY_PROTECTED_LOCAL_BYPASS_GROUPS = frozenset(
    (SERVICE_DISCORD, SERVICE_YOUTUBE)
)

RECOVERY_NONE = "none"
RECOVERY_INVALIDATE_STRATEGY = "invalidate_strategy"
RECOVERY_RESWEEP_EXACT_HOST = "resweep_exact_host"
RECOVERY_RESTART_OWNED_GEPH = "restart_owned_geph"
RECOVERY_RECHECK = "recheck"
RECOVERY_WARN_EXTERNAL = "warn_external"


@dataclass(frozen=True)
class ConnectionOutcome:
    host: str
    service_group: str
    route_class: str
    backend: str
    failure_phase: str
    bytes_received: int
    duration: float
    reason: str
    ok: bool


@dataclass(frozen=True)
class RecoveryContext:
    backend_owned: bool = False
    restart_recommended: bool = False
    restart_rate_limited: bool = False
    strategy_invalidation_recommended: bool = False
    recheck_recommended: bool = False
    external_state: bool = False


@dataclass(frozen=True)
class RecoveryAction:
    kind: str
    target: str = ""
    reason: str = ""


def reduce_connection_outcome(outcome, context=None):
    """Choose safe recovery work without performing any side effects."""
    context = context or RecoveryContext()
    if outcome.ok:
        return (RecoveryAction(RECOVERY_NONE),)

    reason = outcome.reason[:200]
    protected_local = outcome.service_group in POLICY_PROTECTED_LOCAL_BYPASS_GROUPS
    if protected_local or outcome.route_class == ROUTE_LOCAL_BYPASS:
        return (
            RecoveryAction(
                RECOVERY_INVALIDATE_STRATEGY,
                outcome.service_group,
                reason,
            ),
            RecoveryAction(RECOVERY_RESWEEP_EXACT_HOST, outcome.host, reason),
            RecoveryAction(RECOVERY_RECHECK, outcome.service_group, reason),
        )

    if context.external_state:
        return (RecoveryAction(RECOVERY_WARN_EXTERNAL, outcome.backend, reason),)

    if outcome.route_class == ROUTE_GEO_EXIT:
        actions = []
        if context.strategy_invalidation_recommended:
            actions.append(
                RecoveryAction(RECOVERY_INVALIDATE_STRATEGY, outcome.host, reason)
            )
        if (
            context.backend_owned
            and context.restart_recommended
            and not context.restart_rate_limited
        ):
            actions.append(
                RecoveryAction(RECOVERY_RESTART_OWNED_GEPH, outcome.backend, reason)
            )
        else:
            actions.append(
                RecoveryAction(RECOVERY_RECHECK, outcome.service_group, reason)
            )
        return tuple(actions)

    if outcome.route_class == ROUTE_UNKNOWN and context.recheck_recommended:
        return (RecoveryAction(RECOVERY_RECHECK, outcome.host, reason),)

    return (RecoveryAction(RECOVERY_NONE),)

File: crates/slipstream-core/src/routing_recovery.rs
//! Pure recovery reduction for the version 1 recovery contract.

use crate::routing_policy::{RouteClass, ServiceGroup};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct ConnectionOutcome {
    pub host: String,
    pub service_group: ServiceGroup,
    pub route_class: RouteClass,
    pub backend: String,
    pub failure_phase: String,
    pub bytes_received: u64,
    pub duration: f64,
    pub reason: String,
    pub ok: bool,
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct RecoveryContext {
    pub backend_owned: bool,
    pub restart_recommended: bool,
    pub restart_rate_limited: bool,
    pub strategy_invalidation_recommended: bool,
    pub recheck_recommended: bool,
    pub external_state: bool,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RecoveryActionKind {
    None,
    InvalidateStrategy,
    ResweepExactHost,
    RestartOwnedGeph,
    Recheck,
    WarnExternal,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct RecoveryAction {
    pub kind: RecoveryActionKind,
    pub target: String,
    pub reason: String,
}

fn action(
    kind: RecoveryActionKind,
    target: impl Into<String>,
    reason: impl Into<String>,
) -> RecoveryAction {
    RecoveryAction {
        kind,
        target: target.into(),
        reason: reason.into(),
    }
}

fn no_recovery() -> Vec<RecoveryAction> {
    vec![action(RecoveryActionKind::None, "", "")]
}

pub fn reduce_connection_outcome(
    outcome: &ConnectionOutcome,
    context: &RecoveryContext,
) -> Vec<RecoveryAction> {
    if outcome.ok {
        return no_recovery();
    }

    let reason: String = outcome.reason.chars().take(200).collect();
    if outcome.service_group.is_protected_local_bypass()
        || outcome.route_class == RouteClass::LocalBypass
    {
        let group = outcome.service_group.to_string();
        return vec![
            action(
                RecoveryActionKind::InvalidateStrategy,
                group.clone(),
                reason.clone(),
            ),
            action(
                RecoveryActionKind::ResweepExactHost,
                outcome.host.clone(),
                reason.clone(),
            ),
            action(RecoveryActionKind::Recheck, group, reason),
        ];
    }

    if context.external_state {
        return vec![action(
            RecoveryActionKind::WarnExternal,
            outcome.backend.clone(),
            reason,
        )];
    }

    if outcome.route_class == RouteClass::GeoExit {
        let mut actions = Vec::with_capacity(2);
        if context.strategy_invalidation_recommended {
            actions.push(action(
                RecoveryActionKind::InvalidateStrategy,
                outcome.host.clone(),
                reason.clone(),
            ));
        }
        if context.backend_owned && context.restart_recommended && !context.restart_rate_limited {
            actions.push(action(
                RecoveryActionKind::RestartOwnedGeph,
                outcome.backend.clone(),
                reason,
            ));
        } else {
            actions.push(action(
                RecoveryActionKind::Recheck,
                outcome.service_group.to_string(),
                reason,
            ));
        }
        return actions;
    }

    if outcome.route_class == RouteClass::Unknown && context.recheck_recommended {
        return vec![action(
            RecoveryActionKind::Recheck,
            outcome.host.clone(),
            reason,
        )];
    }

    no_recovery()
}
